fused-effects-time/src/Control/Carrier/Time/System.hs

```diff
@@ -0,0 +1,56 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Control.Carrier.Time.System
+( -- * Time carrier
+  Instant(..)
+, Duration(..)
+, since
+, time
+, TimeC(..)
+  -- * Time effect
+, module Control.Effect.Time
+) where
+
+import Control.Algebra
+import Control.Applicative (Alternative)
+import Control.Carrier.Lift
+import Control.Effect.Time
+import Control.Monad (MonadPlus)
+import Control.Monad.Fix (MonadFix)
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.Trans.Class (MonadTrans(..))
+import Data.Fixed
+import Data.Time.Clock.System
+
+newtype Instant = Instant { getInstant :: SystemTime }
+  deriving (Eq, Ord, Show)
+
+newtype Duration = Duration { getDuration :: Nano }
+  deriving (Eq, Fractional, Num, Ord, Real, Show)
+
+
+since :: Instant -> Instant -> Duration
+since (Instant (MkSystemTime bs bns)) (Instant (MkSystemTime as ans)) = Duration (realToFrac (as - bs) + MkFixed (fromIntegral ans - fromIntegral bns))
+{-# INLINABLE since #-}
+
+time :: Has (Time Instant) sig m => m a -> m (Duration, a)
+time = timeWith since
+{-# INLINE time #-}
```

Comment:
Individual carriers can (and should) define instants, durations, and a wrapper for `timeWith`.